### ISSUE TITLE
High percentile of cursor.put show that some calls are very slow

### ISSUE THREAD

**Issue by AskAlexSharov** [erthink/libmdbx#132]
We have next piece of code in app to update DupSort DBI:
```
err := cursor.put(key, value, MDBX_NOOVERWRITE)
if err != nil {
   if mdbx.IsKeyExists(err) {
      return c.put(key, value, MDBX_CURRENT)
   }
   return err
}
```

I built prometheus dashboard for latency of this code:
- 0.5 percentile show very similar speed of lmdb and mdbx - ~1microsecond
- 0.75 percentile show - ~2microseconds vs ~3microseconds
- 0.99 percentile shows - ~200microseconds vs ~2milliseconds 

<img width="1162" alt="Screen Shot 2020-11-03 at 11 53 37" src="https://user-images.githubusercontent.com/46885206/97951308-4ef2c900-1dcc-11eb-8a79-6276fc6b34b3.png">
<img width="1158" alt="Screen Shot 2020-11-03 at 11 53 45" src="https://user-images.githubusercontent.com/46885206/97951313-55814080-1dcc-11eb-8e20-ab08b1ffc86c.png">
<img width="1174" alt="Screen Shot 2020-11-03 at 11 54 13" src="https://user-images.githubusercontent.com/46885206/97951312-54501380-1dcc-11eb-9ca7-a5790baff938.png">

Updates flushing by big sorted batches (we store updates in std:map, then sort, and flush to db in 1 tx).

**Comment by erthink** [comment]
Are you sure you disabled [assertions](https://en.wikipedia.org/wiki/Assert.h) by `- DNDEBUG` while building libmdbx?

**Comment by AskAlexSharov** [comment]
Yes, i using default Gnumake (which i get after make dist) to build mdbx. And i saw today that -NDEBUG=1 is used for building. 

But i have small piece of C code to help with Go bindings - which i building by this flags: https://github.com/ledgerwatch/turbo-geth/blob/1b4f721a614a7628d85e2e24e782e2d07aab9bf8/ethdb/mdbx/mdbx.go#L136

And then link with libmdbx.a - but i’m not sure that linking flags are correct. Which are correct?

**Comment by erthink** [comment]
1. The `CFLAGS` for bindings seem boilerplate (i.e. no problem), but I would prefer to add `-DNDEBUG=1` to a release builds for sameness/uniformity.
2. The ["Quick reference for Insert/Update/Delete operations"](https://erthink.github.io/libmdbx/group__c__crud.html#c_crud_hints) may be useful.
I.e. you may use one call to update instead of by two.
3. With a dynamic database size, there can be large spikes in latency when a database file size increases.
4. Even with a fixed database size, MDBX actively uses `madvise()`, which can also lead to latency spikes, but should ensure better memory usage in the system as a whole.
5. AFAIK, LMDB has some problems when using `MDB_CURRENT` for a DUPSORT-databases. For instance, LMDB may loose data when new data value longer than old, or leave some unused hole in a page if new value is shorter.
I.e. in some conditions MDBX will spend more CPU clocks to avoid such troubles.
6. The current implementation of dirty page list in not ideal, especially for large transactions (see TODO in the ChangeLog).
In short, the current implementation has a cumulative effect, where saving on multiple operations subsequently leads to a spike in latency. The problem is that for large transactions (for a long list of dirty pages), such spike exceeds the overall savings.
I'm planning resolve this TODO after all (or most of) current bugs was fixed.

If you want to dig deeper, you should use ioarena and / or VTune.

**Comment by AskAlexSharov** [comment]
Thank much. 
2. In that place of my app - only way to avoid 2 operations - is using custom comparator. maybe need introduce it here, but historically we don't.
3. I can set growth step to 100Gb and re-test. 
4. Don't know if I told you or not - we use MDBX_NORDAHEAD by default - we have this flag from first day of LMDB benchmarks. 
5. Interesting. 
6. For testing purpose I can make transactions smaller (i also noticed some cumulative effect). 

"you should use ioarena and / or VTune." - will do tomorrow. Feel free to recommend me ways to introspect LMDB/MDBX based apps and OS. 
Golang - is a condition, but we also have C++ version of our app, just i'm on Go side for now.

**Comment by erthink** [comment]
> 4. Don't know if I told you or not - we use MDBX_NORDAHEAD by default - we have this flag from first day of LMDB benchmarks.

This is maybe not good for libmdbx.
Without `MDBX_NORDAHEAD` libmdbx will automatically enable/disable readahead depending of DB size and available RAM.
Therefore, I would prefer that you do not use `MDBX_NORDAHEAD` to evaluate the effectiveness of libmdbx automation.

**Comment by AskAlexSharov** [comment]
Ok, let me run some tests without this flag.

**Comment by AskAlexSharov** [comment]
I made an experiment on ioarena which shows that `througput` of mdbx is higher, but `max latency` is much slower than lmdb (in my understanding `max latency` is a bit similar metric as `high percentile`). 

Settings: master branch of ioarena, Batch, Lazy, small keys (to enforce updates), run multiple time on same `_ioarena` folder to enforce growth of freelist. 

```
for i in {1..5}; do src/ioarena -m lazy -D mdbx -B batch -k 6 -v 5 -n 10000000; done
total:        77.987s              20000
min latency:   2.624ms/op
avg latency:   3.899ms/op
rms latency:  20.783ms/op
max latency: 702.538ms/op
throughput: 255.428 ops/s




for i in {1..5}; do src/ioarena -m lazy -D lmdb -B batch -k 6 -v 5 -n 10000000; done
total:        74.272s              20000
min latency:   3.220ms/op
avg latency:   3.714ms/op
rms latency:   3.716ms/op
max latency:   9.291ms/op
throughput: 268.121 ops/s
```

Branch which modify ioarena to workload closer to our app is here (noreadahead, lifo for lazy ioarena mode, bigger batch size, bigger mapsize to experiment on 20-30GB dataset): 
https://github.com/pmwkaa/ioarena/compare/master...AskAlexSharov:big_batch?expand=1

This modification also shows similar thing: mdbx thougput is bigger, min latency and avg latency smaller, max latency several times bigger than lmdb.

**Comment by erthink** [comment]
Thank for benchmarking and reporting.
As I wrote, this is mostly expected.
I plan to fix this after the v0.9.2 release.

**Comment by AskAlexSharov** [comment]
Nice. 

`As I wrote, this is mostly expected` - It doesn't really match with your comment https://github.com/erthink/libmdbx/issues/132#issuecomment-721197794 - because:
- default ioarena batch size is small - 500 keys (means no big degradation expected)
- ioarena and my app doesn't use MDB_CURRENT
- ioarena has 4GB geometry - and commands above use only 1Gb  (means no resize - or resize is same as LMDB)

**Comment by erthink** [comment]
На всякий случай я поясню на русском почему "this is mostly expected":
- В контексте пишущей транзакции поддерживается такая штука как "dirty page list", т.е. список измененных/грязных страницы БД.
- Важно что в режимах без `MDBX_WRITEMAP` (когда данные БД отображены в память только для чтения) этот список содержит теневые копии измененных страниц БД, и поэтому любое чтение страницы по номеру требует поиска в этом списке.
- Соответственно поиск в "dirty page list" оказывает существенно влияние на производительность.

В LMDB этот список поддерживается в виде отсортированного массива, т.е. при каждом добавлении выполняется двоичный поиск и вставка с раздвижкой элементов. Соответственно, для поиска страниц также всегда используется binary search.

В MDBX для этого списка реализована ленивая сортировка, т.е. при добавлении новый элемент просто дописывается в конец. А поиск выполняется чуть более хитро:
 - если несортированный хвост не большой, то выполняется линейный поиск в этом хвосте, а затем двоичный поиск в отсортированном теле.
- иначе выполняется сортировка с последующим двоичным поиском.
- для мелких транзакций это даёт наблюдаемый выигрыш, но с ростом размера текущая реализация до-сортировки начинает работать в минус.
- проблема решается заменой полной пере-сортировки массива на сортировку хвоста с последующим merge (что и записано в TODO).

**Comment by erthink** [comment]
Ну к описанному выше добавляется еще один важный момент из-за разницы в опциях использованных в ioarena для _lazy_ режима:
- для LMDB: `MDB_NOSYNC | MDB_NOMETASYNC` = страницы перезаписываются и никаких `fdatasync()`, __при аварии можно потерять всю БД__.
- для MDBX: `MDBX_SAFE_NOSYNC | MDBX_NOMETASYNC` (причем `MDBX_LIFORECLAIM` не даёт эффекта из-за `MDBX_SAFE_NOSYNC`, см доку) = страницы НЕ перезаписываются до `fdatasync()`, __при аварии гарантируется сохранение последней транзакции полностью записанной на диск__.

**Comment by AskAlexSharov** [comment]
Интересно - я попробую посмотреть на наши графики если включить WRITE_MAP.

**Comment by erthink** [comment]
> Интересно - я попробую посмотреть на наши графики если включить WRITE_MAP.

Да, но важнее выровнять сравниваемые режимы LMDB и MDBX.

**Comment by AskAlexSharov** [comment]
Для того чтобы очистить список грязных страниц - нужен sync или только commit достаточно?

**Comment by erthink** [comment]
> Для того чтобы очистить список грязных страниц - нужен sync или только commit достаточно?

Коммит и ничего другое

**Comment by AskAlexSharov** [comment]
yep, with WRITE_MAP our app has no problems with flushing std::map into db.

**Comment by AskAlexSharov** [comment]
@erthink , do you need me to run any test on devel?

**Comment by erthink** [comment]
> @erthink , do you need me to run any test on devel?

Not for now. I will notify you.

**Comment by erthink** [comment]
@AskAlexSharov, еще могут быть какие-то правки, но более-менее я закончил всё что хотел сделать.
По-возможности прошу прогнать ваши тесты/бенчмарки на текущей веке `devel`.
Было-бы интересно увидеть разницу в сравнении с предыдущими вашими результатами.

**Comment by AskAlexSharov** [comment]
ок. щас запустим.
